experimental/iterators/test/lib/Conversion/OneToNTypeConversion/OneToNTypeConversion.cpp

```diff
@@ -0,0 +1,322 @@
+// cast happens to be allocated with exactly the same pointer. Dealing with that
+// possiblity is an open TODO.) Also, we do not track which inserted casts are
+// needed for source, target, or argument materialization, so we do some
+// educated guessing to recover that information. Fixing both issues would
+// require to use a PatternRewriter that overloads various `notify*` functions
+// and similar and tracks all changes there. However, that would require a
+// dedicated pattern application driver, which is currently also left as an open
+// TODO.)
+LogicalResult applyOneToNConversion(Operation *op,
+                                    OneToNTypeConverter &typeConverter,
+                                    const FrozenRewritePatternSet &patterns) {
+  // Remember existing unrealized casts.
+  SmallSet<UnrealizedConversionCastOp, 4> existingCasts;
+  op->walk(
+      [&](UnrealizedConversionCastOp castOp) { existingCasts.insert(castOp); });
+
+  // Apply provided conversion patterns.
+  if (failed(applyPatternsAndFoldGreedily(op, patterns)))
+    return failure();
+
+  // Find all newly inserted unrealized casts (that haven't folded away).
+  SmallVector<UnrealizedConversionCastOp> worklist;
+  op->walk([&](UnrealizedConversionCastOp castOp) {
+    if (!existingCasts.contains(castOp))
+      worklist.push_back(castOp);
+  });
+
+  // Replace new casts with user materializations.
+  IRRewriter rewriter(op->getContext());
+  for (UnrealizedConversionCastOp castOp : worklist) {
+    // Create user materialization.
+    TypeRange resultTypes = castOp->getResultTypes();
+    rewriter.setInsertionPoint(castOp);
+    SmallVector<Value> materializedResults;
+
+    // Determine whether operands or results are already legal to know which
+    // kind of materilization this is.
+    ValueRange operands = castOp.getOperands();
+    bool areOperandTypesLegal = llvm::all_of(
+        operands.getTypes(), [&](Type t) { return typeConverter.isLegal(t); });
+    bool areResultsTypesLegal = llvm::all_of(
+        resultTypes, [&](Type t) { return typeConverter.isLegal(t); });
+
+    if (!areOperandTypesLegal && areResultsTypesLegal && operands.size() == 1) {
+      // This is a target materilization.
+      std::optional<SmallVector<Value>> maybeResults =
+          typeConverter.materializeTargetConversion(
+              rewriter, castOp->getLoc(), resultTypes, operands.front());
+      if (!maybeResults)
+        return failure();
+      materializedResults = maybeResults.value();
+    } else if (areOperandTypesLegal && !areResultsTypesLegal &&
+               resultTypes.size() == 1) {
+      // This is a source or an argument materialization.
+      std::optional<Value> maybeResult;
+      if (llvm::all_of(operands, [&](Value v) { return v.getDefiningOp(); })) {
+        // This is an source materialization.
+        maybeResult = typeConverter.materializeArgumentConversion(
```

Comment:
The code doesn't match the comment. Form the check, it appears that the comment is correct and the code is wrong.